ssz/src/main/java/tech/pegasys/teku/ssz/schema/impl/SszUnionSchemaImpl.java

```diff
@@ -0,0 +1,217 @@
+      throw new SszDeserializeException(
+          "Invalid selector " + selector + " for Union schema: " + this);
+    }
+    SszSchema<?> valueSchema = childrenSchemas.get(selector);
+    TreeNode valueNode = valueSchema.sszDeserializeTree(reader);
+
+    return createUnionNode(valueNode, selector);
+  }
+
+  public int getSelectorFromSelectorNode(TreeNode selectorNode) {
+    checkArgument(selectorNode instanceof LeafDataNode, "Invalid selector node");
+    LeafDataNode dataNode = (LeafDataNode) selectorNode;
+    Bytes bytes = dataNode.getData();
+    checkArgument(bytes.size() == SELECTOR_SIZE_BYTES, "Invalid selector node");
+    int selector = bytes.get(0) & 0xFF;
+    checkArgument(selector <= MAX_SELECTOR, "Selector exceeds max value");
+    return selector;
+  }
+
+  public TreeNode getValueNode(TreeNode unionNode) {
+    return unionNode.get(GIndexUtil.LEFT_CHILD_G_INDEX);
+  }
+
+  private TreeNode getSelectorNode(TreeNode unionNode) {
+    return unionNode.get(GIndexUtil.RIGHT_CHILD_G_INDEX);
+  }
+
+  public int getSelector(TreeNode unionNode) {
+    int selector = getSelectorFromSelectorNode(getSelectorNode(unionNode));
+    checkArgument(selector < getTypesCount(), "Selector is out of bounds");
+    return selector;
+  }
+
+  private TreeNode createUnionNode(TreeNode valueNode, int selector) {
+    return BranchNode.create(valueNode, createSelectorNode(selector));
+  }
+
+  @Override
+  public void storeBackingNodes(
+      TreeNodeStore nodeStore, int maxBranchLevelsSkipped, long rootGIndex, TreeNode node) {
+    TreeNode selectorNode = getSelectorNode(node);
+    TreeNode valueNode = getValueNode(node);
+    int selector = getSelectorFromSelectorNode(selectorNode);
+    nodeStore.storeLeafNode(selectorNode, GIndexUtil.gIdxRightGIndex(rootGIndex));
+
+    getChildSchema(selector)
+        .storeBackingNodes(
+            nodeStore, maxBranchLevelsSkipped, GIndexUtil.gIdxLeftGIndex(rootGIndex), valueNode);
+
+    nodeStore.storeBranchNode(
+        node.hashTreeRoot(),
+        rootGIndex,
+        1,
+        new Bytes32[] {valueNode.hashTreeRoot(), selectorNode.hashTreeRoot()});
+  }
+
+  @Override
+  public TreeNode loadBackingNodes(TreeNodeSource nodeSource, Bytes32 rootHash, long rootGIndex) {
+    if (TreeUtil.ZERO_TREES_BY_ROOT.containsKey(rootHash) || rootHash.equals(Bytes32.ZERO)) {
+      return getDefaultTree();
+    }
+
+    final CompressedBranchInfo branchData = nodeSource.loadBranchNode(rootHash, rootGIndex);
+    checkState(
+        branchData.getChildren().length == 2, "Union root node must have exactly two children");
+    checkState(branchData.getDepth() == 1, "Union root node must have depth of 1");
+    final Bytes32 valueHash = branchData.getChildren()[0];
+    final Bytes32 selectorHash = branchData.getChildren()[1];
+    final int selector =
+        nodeSource
+            .loadLeafNode(selectorHash, GIndexUtil.gIdxRightGIndex(rootGIndex))
+            .getInt(0, ByteOrder.LITTLE_ENDIAN);
```

Comment:
Could/should this just use `getSelectorFromSelectorNode(nodeSource.loadLeafNode(selectorHash, GIndexUtil.gIdxRightGIndex(rootGIndex))`?

Reply:
Tried your suggestion, but `nodeSource.loadLeafNode(selectorHash, GIndexUtil.gIdxRightGIndex(rootGIndex))` returns me `0x0100000000000000000000000000000000000000000000000000000000000000` instead of `0x01`. 
Is it a bug or expected behavior? (the actual value of selector leaf node is `0x01`)

Reply:
Ok, leaving it 'as is' for now. At least the same code is in the `SszListSchemaImpl` (borrowed it from there). And probably this needs to be sorted out separately

Reply:
ah yeah of course.  We don't store leaf nodes with data shorter than 32 bytes because the hash is just the data padded out to 32 bytes which means they don't have a unique hash.  So you have to trim the data back down to the right length after loading.  So yeah, the way you have it makes sense (pretty sure I went round this loop doing the list schema code too)...